espressostreamer/espresso_streamer.go

```diff
@@ -0,0 +1,254 @@
+		// that wasn't finalized at all
+		latestBlock, err := s.espressoClient.FetchLatestBlockHeight(ctx)
+		if err != nil {
+			log.Warn("unable to fetch latest hotshot block", "err", err)
+			return err
+		}
+		log.Info("Started node at the latest hotshot block", "block number", latestBlock)
+		s.nextHotshotBlockNum = latestBlock
+	}
+
+	nextHotshotBlockNum := s.nextHotshotBlockNum
+	arbTxns, err := s.espressoClient.FetchTransactionsInBlock(ctx, nextHotshotBlockNum, s.namespace)
+	if err != nil {
+		log.Warn("failed to fetch the transactions", "err", err)
+		return err
+	}
+	if len(arbTxns.Transactions) == 0 {
+		log.Info("No transactions found in the hotshot block", "block number", nextHotshotBlockNum)
+		return nil
+	}
+
+	for _, tx := range arbTxns.Transactions {
+		// Parse hotshot payload
+		attestation, userDataHash, indices, messages, err := arbutil.ParseHotShotPayload(tx)
+		if err != nil {
+			log.Warn("failed to parse hotshot payload, will retry", "err", err)
+			return err
+		}
+		// if attestation verification fails, we should skip this message
+		// Parse the messages
+		if len(userDataHash) != 32 {
+			log.Warn("user data hash is not 32 bytes")
+			continue
+		}
+		userDataHashArr := [32]byte(userDataHash)
+		err = s.verifyAttestationQuote(ctx, attestation, userDataHashArr)
+		if err != nil {
+			log.Warn("failed to verify attestation quote", "err", err)
+			continue
+		}
+		for i, message := range messages {
+			var messageWithMetadata arbostypes.MessageWithMetadata
+			err = rlp.DecodeBytes(message, &messageWithMetadata)
+			if err != nil {
+				log.Warn("failed to decode message, will retry", "err", err)
+				// Instead of returnning an error, we should just skip this message
+				continue
+			}
+			if indices[i] < s.currentMessagePos {
+				log.Warn("message index is less than current message pos, skipping", "messageIndex", indices[i], "currentMessagePos", s.currentMessagePos)
+				continue
+			}
+			// Only append to the slice if the message is not a duplicate
+			if s.findMessageIndexByPos(indices[i]) == -1 {
+				s.messageWithMetadataAndPos = append(s.messageWithMetadataAndPos, &MessageWithMetadataAndPos{
+					MessageWithMeta: messageWithMetadata,
+					Pos:             indices[i],
+					HotshotHeight:   nextHotshotBlockNum,
+				})
+				log.Info("Added message to queue", "message", indices[i])
+			}
+		}
+	}
+
+	// Sort the messagesWithMetadataAndPos based on ascending order
+	// This is to ensure that we process messages in the correct order
+	sort.SliceStable(s.messageWithMetadataAndPos, func(i, j int) bool {
+		return s.messageWithMetadataAndPos[i].Pos < s.messageWithMetadataAndPos[j].Pos
+	})
+
+	return nil
+}
+
+/*
+Check if the message with metadata and pos is already in the queue
+*/
+func (s *EspressoStreamer) findMessageIndexByPos(pos uint64) int {
+	for i, msg := range s.messageWithMetadataAndPos {
+		if msg.Pos == pos {
+			return i
+		}
+	}
+	return -1
```

Comment:
Is this the same as [slice.IndexFunc](https://pkg.go.dev/slices#IndexFunc)? This will be more concise and enhance the readablity

Reply:
@ImJeremyHe yes looks like this function is doing the same thing, feel free to replace it